### DISCUSSION TITLE
[cloud.dify.ai] Node HTTP Request run failed: Reached maximum retries (3) for URL https://api.dify.ai/v1/chat-messages

### DISCUSSION THREAD

**Discussion by posmyk** [langgenius/dify#6968] [Help]
### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

I use Node HTTP Request to connect to other Dify Application. 
There is a POST to: https://api.dify.ai/v1/chat-messages but after first successfull request in first iteration the rest are timed out with the following error: Node HTTP Request  run failed: Reached maximum retries (3) for URL https://api.dify.ai/v1/chat-messages",

### 2. Additional context or comments

Iteration:
![Screenshot 2024-08-05 at 11 46 02](https://github.com/user-attachments/assets/3fa3573a-6ba5-4a5c-b18e-9e547a51003b)

http req:
![Screenshot 2024-08-05 at 11 41 20](https://github.com/user-attachments/assets/c6ae2be8-8c39-47b9-9aca-ab8def21951f)

Test run: Iteration 1 is OK, but the second fails with the timeout:
![Screenshot 2024-08-05 at 11 43 58](https://github.com/user-attachments/assets/ea420c7f-1fee-40b6-a835-269b2f0b7977)

**Comment by MatheusBomtempo** [comment]
Did you find any solution to this problem?

**Comment by zhengjianping1010** [comment]
Did you find any solution to this problem?

**Comment by ccxuy** [comment]
Did you find any solution to this problem?
I tried #6244 , disabled SSRF_PROXY_HTTP_URL, SSRF_PROXY_HTTPS_URL, but still no luck.

**Comment by FrostbiteLee** [comment]
i meet the same problem. 
i want to get the Knowledge Base file list from Node HTTP Request, url is "http://localhost/v1/datasets?page=1&limit=20", but failed.
When i change url to "https://httpbin.org/get", it's ok.

**Comment by Tenold-a** [comment]
I deployed **Dify** on a **Windows 10** system using **Docker Desktop**. Initially, I attempted to send HTTP requests via a **code node**, but it kept failing and say: The requested URL could not be retrieved.  

After checking the documentation, I discovered that **Dify's sandbox environment restricts HTTP requests from the code**.  

### Switching to the HTTP Node  

Since the code node couldn’t send requests, I switched to **Dify's HTTP node** but encountered the same issue as you. I spent a lot of time troubleshooting, trying various fixes, including removing the **SSRF_PROXY** settings mentioned in (https://github.com/langgenius/dify/issues/6244), but nothing worked.  

### The Key

After extensive searching, I know that:  

- **Dify's code execution happens inside a sandbox container.**  
- **The API service is hosted in a separate container.**  
- Containers in Docker can communicate with each other using **container name + port number**, as long as they are on the **same internal network**.  

### The Fix  

I checked the **Nginx configuration** and in [default.conf.template](https://github.com/langgenius/dify/blob/main/docker/nginx/conf.d/default.conf.template) found that requests to `localhost/v1` were actually being forwarded to `http://api:5001`.  


```
    location /v1 {
      proxy_pass http://api:5001;
      include proxy.conf;
    }
```

So, I changed my HTTP request to:  

```http
http://api:5001/v1
```

It is work for me

**Reply by ocyisheng** [reply]
爱老虎油

**Reply by ccxuy** [reply]
@Tenold-a Thanks for sharing this. How to change HTTP request you just mention?

**Reply by WhitePlusMS** [reply]
Changing HTTP to api:5001 solved my problem, thank you ：）

**Comment by ocyisheng** [comment]
我使用本地部署的方式，访问知识库，使用以上方法替换服务器地址，成功访问。在 2025年2月27日，17:01，Misaka Mikoto ***@***.***> 写道：﻿
@Tenold-a Thanks for sharing this. How to change HTTP request you just mention?

**Reply by ccxuy** [reply]
怎么替换呢？是修改default.conf还是dify网页端的界面，我没找到合适的地方改http://api:5001/v1

**Comment by ocyisheng** [comment]
在HTTP节点API参数中，直接使用api:5001替换localhost在 2025年2月28日，10:41，Misaka Mikoto ***@***.***> 写道：﻿
怎么替换呢？是修改default.conf还是dify网页端的界面，我没找到合适的地方改http://api:5001/v1

**Comment by LEON-JU** [comment]
I met this problem while trying to save text to knowledge base through http block, first I validate the curl command using postman and it returns 200 OK:

`curl --location 'http://127.0.0.1/v1/datasets/72f16827-56d6-4296-af48-a844be30b361/document/create-by-text' \
--header 'Authorization: Bearer dataset-VAFeOy91jrWNEBzeaRqTAklQ' \
--header 'Content-Type: application/json' \
--data '{
	"name": "name",
    "text": "text",
    "indexing_technique": "high_quality",
    "process_rule": {"mode": "automatic"}
}'`

but when i switch to the http block, it keeps returning Reached maximum retries.

You may want to replace http://127.0.0.1/v1 with http://host.docker.internal/v1. It works well for me, I guess this is due to some container access issue. Please let me know if anyone has more question about this issue
